### ISSUE TITLE
[feature request] Allow customization of the filename of saved articles

### ISSUE THREAD

**Issue by emanruse** [newsboat/newsboat#1417]
Please allow customization of the filename of saved articles. Example config line:
```
article-save-filename-format "%feedtitle/%D-%t.txt"
article-save-filename-word-separator "-"
```
would result in a filepath prefixed by date and with dashes (instead of underscores):
```
some-feed-title/2021-01-03-some-feed-title.txt
```
This would be very useful when doing a save-all. Currently I do it using a custom script which reads the "Date: ..." line and renames the text file accordingly.

**Comment by Minoru** [comment]
Yup, I think this makes sense, although I'd implement it a bit differently: Podboat has `download-path` and `download-filename-format`, and Newsboat already has `save-path`, so I'd introduce `save-filename-format`. That new setting would have some placeholder (say, `%o`) that is replaced by what Newsboat generates today, but would also have placeholders for title, author and so on. Unlike `%o`, those other placeholders won't be processed in any way, i.e. spaces won't be replaced by underscores. If a user wants that, they can rename the files afterwards with rename(1p) or a similar tool. Does that make sense, @emanruse?

**Comment by emanruse** [comment]
> [...] I'd introduce `save-filename-format`.

OK.

> That new setting would have some placeholder (say, `%o`) that is replaced by what Newsboat generates today, but would also have placeholders for title, author and so on.

Will "so on" also allow something like YYYY-DD-MM-HH-MM-SS prefix for example?

> Unlike `%o`, those other placeholders won't be processed in any way, i.e. spaces won't be replaced by underscores. If a user wants that, they can rename the files afterwards with rename(1p) or a similar tool. Does that make sense, @emanruse?

I am not sure I understand. What is "%o"? Are you suggesting that the user should run rename to replace spaces with another character?

Personally I would like to have my files named as:

Original title: Some example title 1234
Article date: 2020-05-03 22:33:44
Filename: 2020-05-03-22-33-44-some-example-title-1234.txt

Someone else may want something else though. Hence the feature request.

**Comment by Minoru** [comment]
> Will "so on" also allow something like YYYY-DD-MM-HH-MM-SS prefix for example?

Good question! Yes, I think we'd go the same route as `download-filename-format`, and introduce [`%F`](https://newsboat.org/releases/2.22.1/docs/newsboat.html#download-filename-format-F), `%H`, `%M`, `%S` and so on. Your example would be written `%Y-%d-%m-%H-%M-%S`.

> What is "%o"? Are you suggesting that the user should run rename to replace spaces with another character?

`%o` is a made-up placeholder for the "suggested filename" that Newsboat already generates today. In other words, if you wanted the behaviour that exists today, you'd just set `save-filename-format "%o"` — indeed, this would be the default.

Yes, if user wants to replace spaces with something else, they will have to run some other program to rename the files. As your example demonstrates, you want not just to replace spaces with dashes, but also put it all into lowercase. And as you point out, someone else may want something else. Dedicated programs already exist to fulfil these post-processing needs, so instead of re-implementing them, Newsboat will just put all the required info into the filename, and let the user re-format that as they see fit.

**Comment by emanruse** [comment]
OK, sounds good as a whole. I am for the Unix approach "do one thing, do it well". However, from a general user viewpoint, it makes sense to convert filesystem unfriendly characters to fs friendly ones. Although spaces are generally accepted by modern filesystems, they are still not a good naming approach and titles do have them, almost always. Having "&" or " -" (note the leading space) in a filename might introduce a further risk if the user pastes the filename in a command containing the filename. IOW - think security.

What do you say?

**Comment by Minoru** [comment]
Filenames of saved articles are already restricted to alpha-numeric characters, which excludes `&`, `-`, and space. Consider this part of your request fulfilled :) However, https://github.com/newsboat/newsboat/issues/1110 argues for this restriction to be lifted. Similarly, https://github.com/newsboat/newsboat/issues/1261 asks for Podboat to restrict the filenames of podcasts.

So I guess I'll pedal back ju-u-ust a little bit and say that there will be a setting that, when enabled, restricts the filename to alpha-numeric ASCII characters. When disabled, it won't affect the filename in any way. This will be tracked in #1110.

**Comment by emanruse** [comment]
Sounds good.
